shared/runtime/gchelper: Add RISC-V RV64I native gchelper.

agatti opened the pull request:
### Summary

This adds native gchelper support to RV64, given that said platform is now under CI and the code changes required are minimal.  The Unix port already enables platform-specific gchelper code (albeit the generic version), so the RV64 platform deserved its own entry in there.

### Testing

This has been tested on the CI environment both using the generic gchelper code and using the assembler version with this change in the Unix port makefile:

```diff
diff --git a/ports/unix/Makefile b/ports/unix/Makefile
index d5bd6d409..2fab170bf 100644
--- a/ports/unix/Makefile
+++ b/ports/unix/Makefile
@@ -214,16 +214,19 @@ SRC_C += \
        $(wildcard $(VARIANT_DIR)/*.c)
 
 SHARED_SRC_C += $(addprefix shared/,\
-       runtime/gchelper_generic.c \
+       runtime/gchelper_native.c \
        timeutils/timeutils.c \
        $(SHARED_SRC_C_EXTRA) \
        )
 
+SRC_O += shared/runtime/gchelper_rv64i.o
 SRC_CXX += \
 
 OBJ = $(PY_O)
 OBJ += $(addprefix $(BUILD)/, $(SRC_C:.c=.o))
 OBJ += $(addprefix $(BUILD)/, $(SRC_CXX:.cpp=.o))
+OBJ += $(addprefix $(BUILD)/, $(SRC_O:.s=.o))
 OBJ += $(addprefix $(BUILD)/, $(SHARED_SRC_C:.c=.o))
```

### Trade-offs and Alternatives

On RV64, a setjmp frame consumes close to half a kilobyte, so having an alternative implementation for gchelper on that platform would give some interesting memory savings.  The proposed implementation takes around 128 bytes of memory, which is far less than the setjmp version.

agatti committed:
shared/runtime/gchelper: Add RISC-V RV64I native gchelper.

Add native gchelper support for 64 bits RISC-V RV64I targets.

Now that RV64 is under CI, this also enables platform-specific ghelper
in the Unix port.

Also changes the data type holding the register contents to something more
appropriate, so in the remote eventuality somebody wants to use this with
RV128 all they have to do is update the `__riscv_xlen` check.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented on ports/unix/mpconfigport.h:
```diff
@@ -105,7 +105,7 @@ typedef long mp_off_t;
 #define MICROPY_ENABLE_GC           (1)
 
-#if !(defined(MICROPY_GCREGS_SETJMP) || defined(__x86_64__) || defined(__i386__) || defined(__thumb2__) || defined(__thumb__) || defined(__arm__))
+#if !(defined(MICROPY_GCREGS_SETJMP) || defined(__x86_64__) || defined(__i386__) || defined(__thumb2__) || defined(__thumb__) || defined(__arm__) || (defined(__riscv) && defined(__riscv_xlen) && (__riscv_xlen == 64)))
```
Does this need the `__riscv_xlen` part of the logic test?  If you build the unix port on RV32 then it should also work with the native NLR.

agatti replied:
You are indeed correct, however consider that having Linux on RV32 is really uncommon.  I don't think I've ever seen an RV32 chip that can run a Linux environment for real work, short of a soft-core running on a high-end FPGA.

If you don't mind I'd leave this here and then provide a PR that handles both review points and brings RV32 under CI.

dpgeorge replied:
OK.

dpgeorge commented on shared/runtime/gchelper.h:
```diff
@@ -41,7 +41,7 @@ typedef uintptr_t gc_helper_regs_t[4];
 #elif defined(__aarch64__)
 typedef uintptr_t gc_helper_regs_t[11]; // x19-x29
-#elif defined(__riscv) && defined(__riscv_xlen) && (__riscv_xlen == 32)
+#elif defined(__riscv) && defined(__riscv_xlen) && ((__riscv_xlen == 32) || (__riscv_xlen == 64))
```
Is the `__riscv_xlen` logic needed at all?  Is `#elif defined(__riscv)` enough?  (Is there any other RV arch other than 32 and 64 bits?)

agatti replied:
There's also [RV128](https://github.com/riscv/riscv-isa-manual/blob/main/src/rv128.adoc), which as you can imagine has 128-bits wide registers.  Can't say I've ever seen that in the wild though.

dpgeorge replied:
Ah, RV128!  Wow.

I guess the check could then be `__riscv_xlen <= 64`, just to make it a tiny bit simpler.

agatti replied:
But what if somebody comes up with [RV16](https://dlnext.acm.org/doi/10.1007/s11390-022-0910-x)?  Jokes aside, `__riscv_xlen <= 64` can work, and according to the [RISC-V C API document](https://github.com/riscv-non-isa/riscv-c-api-doc/blob/main/riscv-c-api.md), `__riscv_xlen` must always be defined.  I'll change this bit, thanks for reminding me.

dpgeorge commented:
As a side note: do you think it's possible to add CI to build and run tests on unix with riscv32?

agatti commented:
> As a side note: do you think it's possible to add CI to build and run tests on unix with riscv32?

In theory it should just be a matter of changing the arch in the GCC command line, as I guess the `libffi` package brought in to build RV64 would handle RV32 as well.  As mentioned in the first review point, I'd prefer sending a PR with just the RV32 bits in, code changes and all.

dpgeorge commented:
> As mentioned in the first review point, I'd prefer sending a PR with just the RV32 bits in, code changes and all.

For sure, a separate PR would be good for that.

(If needed you can disable ffi altogether for the unix RV32 CI.)

agatti commented:
I've cleaned up the `__riscv_xlen` bits so all currently testable configurations are covered.  I've also changed the data type holding the register contents to something more appropriate, so in the remote eventuality somebody wants to use this with RV128 all they have to do is update the `__riscv_xlen` check.

agatti commented:
Seems like the RISC-V Linux compiler toolchain that's available in the CI environment cannot be used to build 32-bits binaries.  GCC itself claims it's multilib-aware but there's no multilib compiler package from the repo, and no 32-bits include/libs are installed with the compiler.

dpgeorge commented:
> Seems like the RISC-V Linux compiler toolchain that's available in the CI environment cannot be used to build 32-bits binaries. GCC itself claims it's multilib-aware but there's no multilib compiler package from the repo, and no 32-bits include/libs are installed with the compiler.

OK, so let's abandon that idea (unix RV32 CI).

Instead I'm getting closer to having the qemu-* ports run native .mpy files as part of CI.  That will help test more of the native emitter.

agatti commented:
> Instead I'm getting closer to having the qemu-* ports run native .mpy files as part of CI. That will help test more of the native emitter.

That'd be great to have.  I'm also working on natmod support for RV32 and being able to build and run .mpy files in QEMU would help a lot (right now only a few natmods from the `examples/` directory work out of the box, but some show similar issues to the failed RV32 tests: wrong parameter types, crashes during init, etc.).
